src/main/java/co/cask/gcp/datastore/sink/RecordToEntityTransformer.java

```diff
@@ -268,77 +272,107 @@ private Key transformKeyLiteralToKey(Schema.Field field, KeyFactory keyFactory,
           .build();
-        break;
       case INT:
       case LONG:
-        Number longValue = record.get(fieldName);
-        outValue = LongValue.newBuilder(longValue.longValue())
+        Number longValue = getValue(record::get, fieldName, fieldType.toString(), Number.class);
+        return LongValue.newBuilder(longValue.longValue())
           .setExcludeFromIndexes(excludeFromIndex)
           .build();
-        break;
       case FLOAT:
       case DOUBLE:
-        Number doubleValue = record.get(fieldName);
-        outValue = DoubleValue.newBuilder(doubleValue.doubleValue())
+        Number doubleValue = getValue(record::get, fieldName, fieldType.toString(), Number.class);
+        return DoubleValue.newBuilder(doubleValue.doubleValue())
           .setExcludeFromIndexes(excludeFromIndex)
           .build();
-        break;
       case BOOLEAN:
-        outValue = BooleanValue.newBuilder(record.get(fieldName))
+        Boolean booleanValue = getValue(record::get, fieldName, fieldType.toString(), Boolean.class);
+        return BooleanValue.newBuilder(booleanValue)
           .setExcludeFromIndexes(excludeFromIndex)
           .build();
-        break;
       case BYTES:
-        byte[] byteArray = record.get(fieldName);
-        outValue = BlobValue.newBuilder(Blob.copyFrom(byteArray))
+        byte[] byteArray = getValue(record::get, fieldName, fieldType.toString(), byte[].class);
+        return BlobValue.newBuilder(Blob.copyFrom(byteArray))
           .setExcludeFromIndexes(excludeFromIndex)
           .build();
-        break;
       case RECORD:
-        StructuredRecord nestedRecord = record.get(fieldName);
+        StructuredRecord nestedRecord = getValue(record::get, fieldName, fieldType.toString(), StructuredRecord.class);
         FullEntity.Builder<IncompleteKey> nestedBuilder = FullEntity.newBuilder();
         Objects.requireNonNull(fieldSchema.getFields(), "Nested Schema fields cannot be empty").forEach(
           nestedField -> nestedBuilder.set(nestedField.getName(),
                                            convertToValue(nestedField.getName(), nestedField.getSchema(),
                                                           nestedRecord, excludeFromIndex)));
-        outValue = EntityValue.newBuilder(nestedBuilder.build())
+        return EntityValue.newBuilder(nestedBuilder.build())
           .setExcludeFromIndexes(excludeFromIndex)
           .build();
-        break;
+      case ARRAY:
+        Schema elementSchema = Schema.recordOf("arrayElementSchema",
+                                               Schema.Field.of(fieldName, fieldSchema.getComponentSchema()
+                                               ));
+        List<?> arrayValues = getValue(record::get, fieldName, fieldType.toString(), List.class);
+        List<Value<?>> values = arrayValues.stream()
+          .map(value -> {
+            StructuredRecord structuredRecord = StructuredRecord.builder(elementSchema)
```

Comment:
It looks like a StructuredRecord is only being created so that convertToValue() can be called? Is it possible to refactor things so that the method looks something like:

`convertToValue(Schema.Field field, Object value, boolean excludeFromIndex)`

That way you can just convert the array values directly instead of wrapping it in a StructuredRecord

Reply:
There is no way to provide Object value instead of record without duplicate checks because we use two different StructuredRecord methods based on schema types:
1. `record.get(fieldName)` for most of the values
2. `record.getTimestamp(fieldName)` for timestamp values

Reply:
Hm I see. This is a deficiency in the design of these logical types in StructuredRecord. Can you leave a comment here then?

Reply:
Done.